### PR TITLE
perf: Use shared drawing context

### PR DESCRIPTION
**Pull request by FD-**
This implements improvement 2 discussed in issue https://github.com/tidbyt/pixlet/issues/350.

The core changes are:

- Create a single graphics context per frame and pass it to all widgets that are to be drawn
- Add another function to the widget interface for calculating paint bounds without actually drawing anything

# Testing
The changes introduced here have the potential to alter the appearance of existing applets. Of course I ensured that all existing tests succeed, but it doesn't seem like they have 100% coverage.

In addition to the existing tests, I wrote a small utility program (not in the commit) that runs pixlet with and without this change on all applets in the community repo. I then compared file sizes to identify applets that did not render identical, and proceeded to investigate on those. Besides widgets involving randomness, I found a few differences in rounding and clipping that I was mostly able to fix.

The only remaining _visual_ difference affects a corner case of the marquee widget: If the scrolled content is longer than the arbitrarily chosen `10*Marquee.height/width`, the text used to be clipped at the end of that boundary (actually quite ugly), while the new implementation does not do that (it just stops scrolling beyond that boundary). This change seems to only affect the mlb_leaders applet.

For completeness, the following applets rendered visually identical, but differed in webp file size by a few bytes: analog_clock, mbta_new_trains, mbta, mlb_scores, mls_scores, nfl_scores, nhl_scores, powerball, tcat_bus_arrivals, verge_taglines, moretransit, ogs_games_viewer, weather_map, ct_lightrail, sound_transit. The reason is likely slightly changed anti-aliasing behaviour (also see the test in `render/animation/transformation_test.go`)

It is also worth noting that I changed the surface drawing color of the plot widget. The logic used there might draw pixels multiple times, which leads to visual artefacts when using alpha blending. To mitigate the issue, I changed the code to derive the fill color by dampening the line color instead of making it translucent. AFAICT, this doesn't visibly affect any existing apps (they were all previously rendering the plot widget on top of a black background anyway).

Lastly, I also adapted the starfield widget, but the changes remain untested. I couldn't find any documentation or reference to the widget anywhere.

# Performance
I also used my utility tool described above for measuring the speedup across all community apps.

Depending on the run, I have consistently seen an average speedup between 2-3x (never below 2.2x, never above 2.9x) on the MacBook Pro I used for development. I did not execute the extensive test on a Raspberry Pi Zero, but see a very similar speedup on the subset of applets I tested there as well.

It is worth noting that more rendering speedup is likely possible, but not while retaining the clipping behaviour of the old implementation.

### CODE REVIEW

**Comment by FD-** [comment]
Sorry for the force-pushes. I had forgotten the tests in render/animation. All should be fine now.

**Comment by FD-** [comment]
It's worth mentioning that this PR relies on https://github.com/tidbyt/gg/pull/2 being merged in. Thus the reason the CI tests are currently failing.

**Comment by FD-** [comment]
Also, I had not adapted the tests in runtime/render. Done now and confirmed that `make test` succeeds locally for me.

File: render/animation/transformation_test.go [hunk 1]
```diff
@@ -264,14 +264,16 @@ func TestTransformationAll(t *testing.T) {
 	// Checking colors with anti-aliasing (when scaling) is more complex.
 	ic := render.ImageChecker{Palette: map[string]color.RGBA{
 		"█": {0xff, 0xff, 0xff, 0xff},
-		"▓": {0x80, 0x80, 0x80, 0x80},
-		"▒": {0x7f, 0x7f, 0x7f, 0x7f},
-		"░": {0x40, 0x40, 0x40, 0x40},
+		"▓": {0x40, 0x40, 0x40, 0x40},
```
Comment (betterengineering) [inline render/animation/transformation_test.go:267]: Any reason the tests in this file have changed?

File: render/plot_test.go [hunk 2]
```diff
@@ -391,17 +391,17 @@ func TestPlotInvertedColor(t *testing.T) {
...
 
 func TestPlotSurfaceFill(t *testing.T) {
 	ic := ImageChecker{
 		Palette: map[string]color.RGBA{
 			"1": {0x0, 0xff, 0x0, 0xff},
-			",": {0x0, 0xff, 0x0, 0x55},
+			",": {0x0, 0x55, 0x0, 0xff},
 			"2": {0xff, 0, 0, 0xff},
-			":": {0xff, 0, 0, 0x55},
+			":": {0x55, 0, 0, 0xff},
```
Comment (betterengineering) [inline render/plot_test.go:404]: Any reason this test had to change?

File: render/widget.go [hunk 3]
```diff
@@ -2,11 +2,15 @@ package render
 
 import (
 	"image"
+
+	"github.com/fogleman/gg"
 )
 
 // A Widget is a self-contained object that can render itself as an image.
 type Widget interface {
```
Comment (betterengineering) [inline render/widget.go:10]: Seems sane 👌

**Review by betterengineering — APPROVED** [review]
Wow! Thanks so much for putting this together! We spent some time looking this over and it looks great after a first pass. There are a few test cases that changed that don't make sense to me after a quick glance - any reason they needed to change?

I'm going to get this merged awhile so we can run similiar tests as yours against our internal apps. Will cut a release or follow up with changes. Thanks again for this change 🙌

File: render/widget.go (hunk 3 above)
Comment (FD-) [reply render/widget.go:10]: I'm not sure I can follow your comment?

**Pull request merged**